d3-bar-chart-race/src/App.js

```diff
@@ -0,0 +1,368 @@
+            bar
+              .transition(transition)
+              .attr("transform", (d) => `translate(${x(d.value)},${y(d.rank)})`)
+              .call((g) =>
+                g
+                  .select("tspan")
+                  .tween("text", (d) =>
+                    textTween((prev.get(d) || d).value, d.value)
+                  )
+              )
+          ));
+    },
+    [n, prev, next, textTween, x, y]
+  );
+
+  // draw axis
+  const axis = useCallback(
+    (svg) => {
+      const g = svg.append("g").attr("transform", `translate(0,${margin.top})`);
+      const axis = d3
+        .axisTop(x)
+        .ticks(width / 150)
+        .tickSizeOuter(0)
+        .tickSizeInner(-barSize * (n + y.padding()));
+
+      return (_, transition) => {
+        g.transition(transition).call(axis);
+        g.select(".tick:first-of-type text").remove();
+        g.selectAll(".tick:not(:first-of-type) line").attr("stroke", "white");
+        g.select(".domain").remove();
+      };
+    },
+    [margin, x, y, n]
+  );
+
+  // format date
+  function formatGroupingDate() {
+    const columnName = columns[config.date]["name"];
+    let formatDate;
+    if (columnName.indexOf("Year") !== -1) {
+      formatDate = d3.timeFormat("%Y");
+    } else if (
+      columnName.indexOf("Month") !== -1 ||
+      columnName.indexOf("Quarter") !== -1
+    ) {
+      formatDate = d3.timeFormat("%Y-%m");
+    } else {
+      formatDate = d3.timeFormat("%Y-%m-%d");
+    }
+    return formatDate;
+  }
+
+  // draw date ticker
+  const ticker = useCallback(
+    (svg) => {
+      if (keyframes.length) {
+        const now = svg
+          .append("text")
+          .style("font", `bold ${barSize}px sans-serif`)
+          .style("font-variant-numeric", "tabular-nums")
+          .attr("text-anchor", "end")
+          .attr("x", width - 30)
+          .attr("y", margin.top + barSize * (n - 0.45))
+          .attr("dy", "0.32em")
+          .text(formatGroupingDate()(keyframes[0][0]));
+
+        return ([date], transition) => {
+          transition.end().then(() => now.text(formatGroupingDate()(date)));
+        };
+      }
+    },
+    [margin, keyframes, n]
+  );
+
+  // create chart
+  function* myGenerator() {
+    const svg = d3.select(ref).attr("viewBox", [0, 0, width, height]);
+    svg.selectAll("*").remove();
+
+    const updateBars = bars(svg);
+    const updateAxis = axis(svg);
+    const updateLabels = labels(svg);
+    const updateTicker = ticker(svg);
+
+    for (const keyframe of keyframes) {
+      const transition = svg
+        .transition()
+        .duration(duration)
+        .ease(d3.easeLinear);
+
+      // Extract the top bar’s value.
+      x.domain([0, keyframe[1][0].value]);
+
+      updateAxis(keyframe, transition);
+      updateBars(keyframe, transition);
+      updateLabels(keyframe, transition);
+      updateTicker(keyframe, transition);
+      yield transition.end();
+    }
+  }
+
+  const iter = myGenerator();
```

Comment:
you're creating a new iterator on every render here; is this expected?

Reply:
No. I have refactored the code and you can take a look. Thanks.